### PR TITLE
1049 - Reference test tool

### PR DESCRIPTION
**Pull request by bradbown**
Implements the first PR for issue #1049 
Updates templates to extend the tests with a `TestWatcher`, the test watcher intercepts failed tests and records to an output file a mapping of the failed test to its module and constraint pair.

### Example
```json
[
    {
        "moduleName": "blockdata",
        "constraints": {
            "value-constraints": [
                "callcodecallcodecall_110_d0g0v0_London[London]",
                "callcodecallcode_11_d0g0v0_London[London]",
                "callcodecallcode_11_SuicideEnd_d0g0v0_London[London]",
                "callcodecallcode_11_OOGE_d0g0v0_London[London]",
                "exp_d0g0v0_London[London]"
            ]
        }
    },
    {
        "moduleName": "txndata",
        "constraints": {
            "txndata-into-blockdata": [
                "callcodecallcodecall_110_d0g0v0_London[London]",
                "callcodecallcode_11_d0g0v0_London[London]",
                "callcodecallcode_11_SuicideEnd_d0g0v0_London[London]",
                "callcodecallcode_11_OOGE_d0g0v0_London[London]",
                "exp_d0g0v0_London[London]"
            ]
        }
    }
]
```

### CODE REVIEW

File: reference-tests/src/test/resources/templates/BlockchainReferenceTest.java.template
Comment (lorenzogentile404) [inline reference-tests/src/test/resources/templates/BlockchainReferenceTest.java.template:1]: These tests classes may be generated automatically. Could you check it to be sure that the annotation for the `TestWatcher` is permanent and does not disappear whenever we re-generate them on a new machine?

File: reference-tests/src/main/java/net/consensys/linea/MapFailedReferenceTestsTool.java [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+
+  public static List<ModuleToConstraints> getModuleToConstraints(
+      String jsonString, JsonConverter jsonConverter) {
+    List<ModuleToConstraints> constraintToFailingTests = new ArrayList<>();
+    if (!jsonString.isEmpty()) {
+      constraintToFailingTests =
+          new ArrayList<>(
+              Arrays.asList(jsonConverter.fromJson(jsonString, ModuleToConstraints[].class)));
+    }
+    return constraintToFailingTests;
```
Comment (lorenzogentile404) [inline reference-tests/src/main/java/net/consensys/linea/MapFailedReferenceTestsTool.java]: the name of the method seems inconsistent with the returned value. Wouldn't it be cleared `getConstraintsToFailingTests`?

File: reference-tests/src/main/java/net/consensys/linea/MapFailedReferenceTestsTool.java [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+        getModuleToConstraints(jsonString, jsonConverter);
+
+    mapFailedConstraintsToTestsToModule(failedConstraints, constraintToFailingTests, testName);
+
+    jsonString = jsonConverter.toJson(constraintToFailingTests);
+    writeToJsonFile(jsonString, jsonOutputPath);
+  }
+
+  private static void mapFailedConstraintsToTestsToModule(
+      Set<String> failedConstraints,
```
Comment (lorenzogentile404) [inline reference-tests/src/main/java/net/consensys/linea/MapFailedReferenceTestsTool.java:57]: are you creating here the different parts of the generated JSON? The name of the method does not look self-explanatory. I suggest to consider adding some comments or splitting the method into smaller ones.

**Review by lorenzogentile404 — CHANGES_REQUESTED** [review]

File: reference-tests/src/test/resources/templates/BlockchainReferenceTest.java.template
Comment (bradbown) [reply reference-tests/src/test/resources/templates/BlockchainReferenceTest.java.template:1]: The `genereateReferenceBlockchainTest` gradle task uses this template to generate the these tests though right? Unless there is another way to generate these too?
If this is the only way then deleting the generated directory and rerunning this task regenerates tests with the `@ExtendWith(ReferenceTestWatcher.class)` again.

File: reference-tests/src/main/java/net/consensys/linea/MapFailedReferenceTestsTool.java (hunk 1 above)
Comment (bradbown) [reply reference-tests/src/main/java/net/consensys/linea/MapFailedReferenceTestsTool.java]: I've renamed it to `getModuleToConstraints()` as it's responsible for converting the json string to a `List<ModuleToConstraints>` or if the string is empty then creating a new empty `List<ModuleToConstraints>`, renamed the corresponding variables which call this method too.

File: reference-tests/src/main/java/net/consensys/linea/MapFailedReferenceTestsTool.java (hunk 2 above)
Comment (bradbown) [reply reference-tests/src/main/java/net/consensys/linea/MapFailedReferenceTestsTool.java:57]: Yes this is the main method for mapping the failed tests, constraint and its parent module together.
I've split the method into smaller ones, so hopefully the method names help describe what's going on at each stage better now.

**Review by lorenzogentile404 — APPROVED** [review]

**Pull request merged**